include/ginkgo/core/base/mpi.hpp

```diff
@@ -0,0 +1,1221 @@
+GKO_REGISTER_MPI_TYPE(std::complex<double>, MPI_C_DOUBLE_COMPLEX);
+
+
+template <typename T>
+inline const T* in_place()
+{
+    return reinterpret_cast<const T*>(MPI_IN_PLACE);
+}
+
+
+/**
+ * This enum specifies the threading type to be used when creating an MPI
+ * environment.
+ */
+enum class thread_type {
+    serialized = MPI_THREAD_SERIALIZED,
+    funneled = MPI_THREAD_FUNNELED,
+    single = MPI_THREAD_SINGLE,
+    multiple = MPI_THREAD_MULTIPLE
+};
+
+
+/**
+ * Class that sets up and finalizes the MPI environment. This class is a simple
+ * RAII wrapper to MPI_Init and MPI_Finalize.
+ *
+ * MPI_Init must have been called before calling any MPI functions.
+ *
+ * @note If MPI_Init has already been called, then this class should not be
+ * used.
+ */
+class environment {
+public:
+    static bool is_finalized()
+    {
+        int flag = 0;
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Finalized(&flag));
+        return flag;
+    }
+
+    static bool is_initialized()
+    {
+        int flag = 0;
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Initialized(&flag));
+        return flag;
+    }
+
+    /**
+     * Return the provided thread support.
+     *
+     * @return the provided thread support
+     */
+    int get_provided_thread_support() { return provided_thread_support_; }
+
+    /**
+     * Call MPI_Init_thread and initialize the MPI environment
+     *
+     * @param argc  the number of arguments to the main function.
+     * @param argv  the arguments provided to the main function.
+     * @param thread_t  the type of threading for initialization. See
+     *                  @thread_type
+     */
+    environment(int& argc, char**& argv,
+                const thread_type thread_t = thread_type::serialized)
+    {
+        this->required_thread_support_ = static_cast<int>(thread_t);
+        GKO_ASSERT_NO_MPI_ERRORS(
+            MPI_Init_thread(&argc, &argv, this->required_thread_support_,
+                            &(this->provided_thread_support_)));
+    }
+
+    /**
+     * Call MPI_Finalize at the end of the scope of this class.
+     */
+    ~environment() { MPI_Finalize(); }
+
+    environment(const environment&) = delete;
+    environment(environment&&) = delete;
+    environment& operator=(const environment&) = delete;
+    environment& operator=(environment&&) = delete;
+
+private:
+    int required_thread_support_;
+    int provided_thread_support_;
+};
+
+
+/**
+ * Returns if GPU aware functionality has been enabled
+ */
+static bool is_gpu_aware()
+{
+#if GKO_HAVE_GPU_AWARE_MPI
+    return true;
+#else
+    return false;
+#endif
+}
+
+
+namespace {
+
+
+/**
+ * A deleter class that calls MPI_Comm_free on the owning MPI_Comm object and
+ * deletes the underlying comm ptr
+ */
+class comm_deleter {
+public:
+    using pointer = MPI_Comm*;
+    void operator()(pointer comm) const
+    {
+        GKO_ASSERT(*comm != MPI_COMM_NULL);
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_free(comm));
```

Comment:
Just a note, not necessarily something we need to change: This throws/aborts inside a destructor, which may not be ideal. MPI_Comm_free probably also checks against MPI_COMM_NULL (considering how eager the MPI impls I checked are to abort on slight issues)
```suggestion
        if (MPI_Comm_free(comm) != MPI_SUCCESS) {
            // something like in CudaExecutor::raw_free
        }
```